monix-tail/shared/src/main/scala/monix/tail/internal/IterantConcat.scala

```diff
@@ -106,31 +107,34 @@ private[tail] object IterantConcat {
     }
   }
 
-  // TODO: Scope is completely broken
+  def concat[F[_], A](lhs: Iterant[F, A], rhs: Iterant[F, A])(implicit F: Sync[F]): Iterant[F, A] =
+    concat(lhs, rhs, _ => F.unit)
+
+  // TODO: Check stack-safety of stack concats
   /**
     * Implementation for `Iterant#++`
     */
-  def concat[F[_], A](lhs: Iterant[F, A], rhs: Iterant[F, A])
+  def concat[F[_], A](lhs: Iterant[F, A], rhs: Iterant[F, A], close: ExitCase[Throwable] => F[Unit])
     (implicit F: Sync[F]): Iterant[F, A] = {
 
     lhs match {
-      case s @ Scope(_, _, _) =>
-        s.runMap(concat(_, rhs))
+      case s @ Scope(_, _, newClose) =>
+        s.runMap(concat(_, rhs, ec => close(ec) >> newClose(ec)))
```

Comment:
Btw, is this stack safe?

Also, I'd rather not have `Scope` the first in the match statement. Normally Scala should generate a  switch for case classes in bytecode, so it doesn't matter much, but the Scala compiler is moody, so I order them by frequency.

I would add `Scope` branches after `Suspend`, but before Last / Halt.

Reply:
Sorry for going off-topic, but how can Scala generate a switch for case classes? 😮 I mean, there's nothing to switch on.